### ISSUE TITLE
[ngrx-ducks] Schematics

### ISSUE THREAD

**Issue by GregOnNet** [co-IT/co-it#5]
## [WIP]

```bash
#                        v custom-dir 1 v duck-name
ng generate duck counter/store/reducer/counter
#                ^ feature-dir ^ custom-dir 2
```

- [ ] should create the file `counter.duck.ts` in `counter/store/reducer/`
- [ ] should look up module file and register Duck in `DucksModule`

**Comment by skydever** [comment]
Answering here ... I am happy if I can help, you can reach out to me of course. 

Currently I run the default `module` schematic first to import the module in a parent module and afterwards I am creating files from my custom blueprints and overwriting default files in the module itself to register the Ducks there etc. - that seems to be an easy solution for that. Maybe the `service` schematic has useful things for that to reach/edit the correct provide block (to do this for the duck only)?

**Comment by skydever** [comment]
Some thoughts on the schematic to add a single `Duck` into an existing module:

* generate a `counter.duck.spec.ts` too (should be compatible with *jest* and *jasmine* (out of the box?), I am using *jest* in my @nrwl/nx workspace - it has the option to choose between *jest* and *jasmine*) and add a corresponding sample action reducer test
* maybe give the option to extend a @ngrx/entity store and add the crud actions (+ tests in spec)? 
  * option to set the action type `[source]` for all of them? Is there a better way to use @ngrx/entity maybe?
    * decorator at class level that gets some options for the action type, like the `[source]` prefix that would create the actions automatically (and the `Duck` methods to dispatch)? no need to test them, the `Duck` is providing it, so no additional tests for this tasks ...
* generate selectors (optional or own schematic maybe, path to create a `from-counter.selector.ts`)
* generate an effect class (optional, path) that imports the `Duck` and adds a corresponding sample effect
  * concerning @ngrx/entity, did not try ngrx data yet. would you recommend it for simple crud operations including the http requests? With a `Duck` all other actions can be as easily added - not sure if using both is a good idea - extending a `Duck` via decorator to add this functionality would be useful to me. Currently I create the actions+effects in my module schematic (optional - I can pass some types of module to my schematic, like store-entity, store, simple ng-module ...) at `Duck`-level via the blueprints.

I did not try any ngrx schematics yet, I used my own, so I am not sure what is already out there that would be helpful too.

What do you think?

**Comment by GregOnNet** [comment]
Hey,

thank you so much for your input.
Please find my thoughts below. 👍 

If I misunderstood something, please let me know.

### Specs

Good idea. Since Jest and Jasmine support the same DSL, it is possible to provide one Template for both.
I wonder which sample test could be beneficial

### @ngrx/entity store & crud actions

This is a good idea as well. I would postpone this feature to the 2nd iteration of the schematics since we need to provide the basic tooling first.

### action type `[soruce]`

I am not sure if I get the idea.
According to the following sample...

```
@Prefix(`[API]`)
class SomeDuck {
  @Action('create')
  create(state: SomeState, payload: SomeModel) {
    /* ... */
  }
}
```

Do you mean that a Decorator (e.g. `@Prefix`) should allow creating action types more easily? 
According to the upper sample the action type for `create()` would be: `[API] create`.

### Selectors

Can you elaborate on your suggestion concerning selectors?
How is this related to the Duck?

For now, I think **ngrx/schematics** does not provide a selector schematic.

### ngrx-data

I tried **ngrx-data** and this library is really handy for pure `CRUD` operations. It works based on a convention setting up Actions, Effects and Reducers.

**ngrx-ducks** simplifies intra-app communication and provides a self-regulatory easy to use Facade.

Both libraries can run in parallel. For now, I do not know how those can be combined or benefit from each other.
I would need to dive deeper into the code base of ngrx-data. 😊

Cheers
Gregor

**Comment by skydever** [comment]
Hey,

I jump right in:

#### Specs

About the sample test, I must admit I do not have that much knowledge on testing - but I really want to increase my knowledge on that - what is useful, when is it a waste of time, when to test what and things like that ... testing reducer functions should be very easy, but do you test every reducer in a *well tested* app? I think the schematic should just show how to create an instance of a `Duck` (just `new MyDuck()`?) and test the sample action that comes with every `Duck` when generated via schematic, like this (if that is how you test a `Duck`? Without much *opinion* in it?):

```typescript
import { MyFeature } from './my-feature.ducks';

describe('MyFeature', () => {
  let duck: MyFeature;
  let state: MyFeatureState;

  beforeEach(() => {
    duck = new MyFeature();
    state = { count: 0 };
  });

  it('should set data', () => {
    expect(duck.set(1).count).toBe(1);
  });
})
```

#### @ngrx/entity store & crud actions

I am just wondering, with basic tooling you mean the setup for a schematic? My current setup is based on a @nrwl/nx workspace schematic, with a little hack to be executed via `ng g`. Are you already working on that or have some plans on how to set this up for the `Ducks`? Where to be located in the repo and things like that? 

The thing is, when I came up with this idea I thought of a decorator for it, and **action type `[source]`** is one of the config values you pass to the decorator - the decorator will create all actions (+effects maybe?) *under the hood* based on some config (eg. prefix, EntityClass). I will create a separate issue to discuss this kind of decorator further. I also thought about a custom base class to extend a `Duck` from to handle this, could be possible too I guess (and initially created via schematics)...

#### **action type `[source]`**

This is actually very funny. As described above my intention was a bit different, but today I spoke with a team mate about this - and he showed me that he was using a custom property (I think outside of the class, but in the same file) to define the source for all actions within the `Duck`, exactly what your `@Prefix` decorator example is showing. Afer a 2nd thought, I am not sure if it is a good idea to do that after all - to set a prefix/source for all of the actions - I am new to really use ngrx and the redux pattern in an app and came across [Angular Checklist](https://angular-checklist.io) and the **capture events with actions, not commands** entry, watched the [Good action hygiene](https://www.youtube.com/watch?v=JmnsEvoy-gY) video by Mike Ryan ... how do you use a `Duck`? Combine actions for 1 source only? Sometimes... 🤔 Maybe the actions, but not the effects - if the *view area* (not to say component) is part of the effect/action type or better, the way of how you define a `[source]` within your project ... my current conclusion is, that is is very easy to create actions using the `Ducks` and I can handle the naming manually, and it would force me to put more thoughts on the type name initially - hopefully :wink: 

#### selectors

You are right, they are not `Duck` specific - but it would be nice to generate them too for an opinionated complete `Duck feature module`, using the right types/name at the `createFeatureSelector()` call ... but this is very structure specific, a complete feature module, and maybe does not belong here after all (?) Writing a selector does not require much basic setup (to zereo), maybe that is why there is none? Could be an addon maybe... 

In my projects I am creating modules with the complete structure via schematics, but I am not able to add a single `Duck` to that module via schematics at the moment, and I think this is most prior case for you too (?) - and I am happy to help with that if I am able to.

#### ngrx-data

I will write about that in #7 (the issue I will create for the @ngrx/entity decorator), and also low to no priority for this issue I think.

thank you for this project, really awesome 👍

**Comment by GregOnNet** [comment]
## Prefix-Decorator

I also think that one Prefix does not match every use case.
I am not sure either if it is possible to access a configured prefix from another decorator. 🤔 

## Selectors

Yes, selectors a playing a big role.
I cannot say what the best way will be, since the Decorator API may allow to inline selectors inside the Duck.
This would be awesome, wouldn't it? 😆

But I am not sure if I can get this to work.

That's why I want to postpone this decision to the point ngrx-ducks 7.3 is released.

## Generation Schematic

In the beginning I want to give the users as much as freedom as possible to create a Duck.
Teams may have different philosophies to structure their ngrx app.

Nevertheless the documentation will provide an opinionated directory structure.

The following sample shows how the "duck schematic" could look like.

```bash
# command
ng generate @co-it/schematics:duck counter-module/counter

# generated structure
counter-module/
├── counter/
    ├── counter.duck.ts
    ├── counter.duck.spec.ts
```

### Parameter

- `--flat` (default: false) create/not create a directory for the Duck
- `--spec` (default: true) create/not create the spec file

## Initial Spec

I like the draft for the unit test of the Duck. 👍

Depending of the final API we may also think about creating an integration test covering the Store as well.

# Thank you

Thanks, for all the input @skydever.
It is awesome.

**Comment by skydever** [comment]
Thx for taking my input into account!

#### Selector

It would be great to get some selectors out of the box from the `Duck`, totally 👍 Sadly I do not know how to help here, my knowledge seems to be a bit limited concerning all this *dynamic magic*. Do you have some resources to share, to get the advanced basics? How to create decorators, what is possible? What are the features of typescript you take advantage the most? Can you recommend something?

#### Schematic

Great starting point! Then it should also register the `Duck` at the nearest `NgModule` if not explicitly specified via `--module`, as you mentioned at the top. This would be great. An opinionated directory structure could be created via `ng generate @co-it/schematics:duck-module counter-module` maybe? This is what I am doing, mostly with blueprint files only. There I can jump right in if you set up the basics, make some proposals (prompts, blueprint files with replacement), as a PR hopefully 😃

**Comment by skydever** [comment]
`ng add @co-it/ngrx-ducks` to setup `package.json` could be useful too

**Comment by GregOnNet** [comment]
## Install library

### Command

`ng add @co-it/ngrx-ducks-schematics`

## Create a Duck

### Command

`ng generate @co-it/ngrx-ducks-schematics:duck <path>/<name>`

### Output

```
# generated structure
<path>/
├── <name>/
    ├── <name>.duck.ts
    ├── <name>.duck.spec.ts
    ├── <name>.duck.state.ts
    ├── index.ts
```

|File|Content|
|----|-------|
|\<name\>.duck.ts|Contains the Duck itself|
|\<name\>.duck.spec.ts|Contains Sample Tests|
|\<name\>.duck.state.ts|Contains the State Interface|

> I will add sample files later on.


## Create selectors

### Command

`ng generate @co-it/ngrx-ducks-schematics:selector <path>/<name>`

**Comment by GregOnNet** [comment]
@skydever I setup the schematics project in the repository.

![image](https://user-images.githubusercontent.com/444278/52766377-4621f880-3027-11e9-8c9d-9772705d88e7.png)

## Cloning the repository

This project is managed by lerna.
To install all dependencies please do the following.

```bash
npm install -g yarn
npm install -g lerna

git clone https://github.com/co-IT/co-it.git
cd co-it

yarn
lerna bootstrap
```

This will download and link all dependencies.
If you have any questions do not hesitate to ask me. 👍

**Comment by skydever** [comment]
great, thank you 👍  I will try to add something!

**Comment by skydever** [comment]
Just to let you know, I will create a draft PR this week.

**Comment by GregOnNet** [comment]
Nice Thank you!

**Comment by skydever** [comment]
I thought about supporting nrwl/nx or angular workspace projects, is there a *best practice* of how to create a schematic to do that? or should the `path` be absolute from the root, no matter what kind of project it is?

... I am kinda blind in terms of what the cli currently supports, I have nrwl/nx custom workspace schematics that wrap the original schematics and I pass the options the way they best fit for my workspace, and they still work (nx workspace) so no changes on what I use so far ... I pass the source starting from `<app-name>/src/app` or `<lib-name>/src/libs` and the project name (via prompts), not sure how a schematic is handling that with an angular workspace ... 
(eg. 
`ng g c my-module/my-component --project=my-app` 
**->** `my-app/src/app/my-module/my-component/my-component.ts|scss|html|spec.ts`
or 
`ng g c my-module/my-component --project=my-lib`
**->** `my-lib/src/lib/my-module/my-component/my-component.ts|scss|html|spec.ts`
). 

Parse `angular.json` check `projects` and `sourceRoot` entries and then decide?

**Comment by GregOnNet** [comment]
It is a challenge to deal with paths, right.
I wrote a helper class that parses the angular.json configuration since you also want to support to generate the files in different projects:

https://github.com/kentan-official/kentan/blob/master/packages/schematics/src/lib/angular-project/angular-cli-six.ts

As far as I know, schematics do not provide an out of the box solution for it.

**Comment by GregOnNet** [comment]
My plan is to ship the schematics directly with ngrx-ducks itself.
In this way, we do not end up with too long commands

```bash
ng g @co-it/ngrx-ducks:duck <path>
# vs
ng g @co-it/ngrx-ducks-schematics:duck <path>
```

**Comment by GregOnNet** [comment]
Schematics have been shipped with version 7.6.0. 🎉 

We need three further improvements:

- [ ] Reducer Function needs to be typed
- [ ] Allow creating a duck without passing `--project`
- [ ] Document all available parameters for the generation command in the README.md

## Reducer Function needs to be typed

Running `ng g @co-it/ngrx-ducks:duck store/search --project ngrx-ducks-demo` produces the duck:

```ts
import { Action, Ducksify, reducerFrom } from '@co-it/ngrx-ducks';
import { SearchState } from './search.state';

@Ducksify<SearchState>({
  initialState: {
    data: 0
  }
})
export class Search {
  @Action('[Search] Sample action type')
  setData(stateSlice: SearchState, data: number): SearchState {
    return { ...stateSlice, data };
  }
}

export function searchReducer(state, action) {
  return reducerFrom(Search)(state, action);
}
```

Having TypeScript's strict mode enabled will break the build since `state` & `action` has the any type.
The `searchReducer` should look like the following:

```ts
export function searchReducer(
  state: SearchState | undefined,
  action: DucksifiedAction
): SearchState {
  return reducerFrom(Counter)(state, action);
}
```

- `state` needs to be a Union Type of `SearchState | undefined` to get the strict mode to work
- `DucksifiedAction` needs to be added to the import statement. It also comes from @co-it/ngrx-ducks

## Allow creating a duck without passing `--project`

If no `--project` is given the first project of the `angular.json` should be taken

@skydever 
- Do you want to take care of this?
- Do you have time for it? :)

**Comment by skydever** [comment]
yes, I can do that :+1:
